tests/test_warp.py

```diff
@@ -1730,6 +1730,34 @@ def test_issue_1446():
     assert round(g["coordinates"][1], 1) == 4212702.1
 
 
+def test_transform_geom_geometry_collection():
+    output = transform_geom(
+        "EPSG:4326",
+        "EPSG:3857",
+        {
+            "type": "GeometryCollection",
+            "geometries": [
+                {
+                    "type": "Point",
+                    "coordinates": [1, 2]
+                }
+            ]
+        }
+    )
+    assert output == {
+        'type': 'GeometryCollection',
+        'geometries': [
+            {
+                'type': 'Point',
+                'coordinates': (
+                    pytest.approx(111319.49079327357),
+                    pytest.approx(222684.20850554405)
+                )
+            }
+        ]
+    }
```

Comment:
can we test https://github.com/rasterio/rasterio/pull/2581/files#r961944737 too 🙏

Reply:
I tried testing it. The issue is that the OGRGeomBuilder has the same error check when initializing the geometry and I haven't figured out how to reach it. It shouldn't be reachable unless the OGRGeomBuilder supports a geometry type that GeomBuilder does not. This is no longer the case with this change from what I can tell. It is mostly a fallback.